pkg/didcomm/protocol/decorator/decorator.go

```diff
@@ -118,7 +126,140 @@ func (d *AttachmentData) Fetch() ([]byte, error) {
 
 	// TODO add support to fetch links
 
-	// TODO add support for jws signatures
-
 	return nil, errors.New("no contents in this attachment")
 }
+
+type rawSig struct {
+	Header    rawSigHeader `json:"header,omitempty"`
+	Signature string       `json:"signature,omitempty"`
+	Protected string       `json:"protected,omitempty"`
+}
+
+type rawSigHeader struct {
+	KID string `json:"kid,omitempty"`
+}
+
+type rawProtected struct {
+	JWK json.RawMessage `json:"jwk,omitempty"`
+	Alg string          `json:"alg,omitempty"`
+}
+
+// Sign signs the base64 payload of the AttachmentData, and adds the signature to the attachment.
+func (d *AttachmentData) Sign(c crypto.Crypto, kh, pub interface{}, pubBytes []byte) error {
+	didKey, _ := fingerprint.CreateDIDKey(pubBytes)
+
+	jwk, err := jose.JWKFromKey(pub)
+	if err != nil {
+		return fmt.Errorf("creating jwk from pub key: %w", err)
+	}
+
+	jwk.Kty = "OKP"
+	jwk.KeyID = didKey
+
+	jwkBytes, err := jwk.MarshalJSON()
+	if err != nil {
+		return fmt.Errorf("marshaling jwk: %w", err)
+	}
+
+	protected := rawProtected{
+		JWK: jwkBytes,
+		Alg: "EdDSA",
+	}
+
+	protectedBytes, err := json.Marshal(protected)
+	if err != nil {
+		return fmt.Errorf("marshaling protected header: %w", err)
+	}
+
+	protectedB64 := base64.RawURLEncoding.EncodeToString(protectedBytes)
+
+	b64data := strings.Trim(d.Base64, "=")
```

Comment:
Manipulating the encoding manually is not recommended.

Prefer to update the code to make the value of `d.Base64` the result of `RawURLEncoding` instead of `StdURLEncoding` and remove the `Trim()` call here.

Reply:
It looks like that would break other protocols that currently use attachments, since they interop using StdEncoding, and [aries-rfcs/#558](https://github.com/hyperledger/aries-rfcs/pull/558) is not yet merged - moreover, that PR states that we should accept _either_ URLEncoding or RawURLEncoding for the data - therefore for signing we'd need to keep this trim, since ACA-Py trims the padding from the attachment base64 data, for a consistent signed payload even if the encoding is different across different uses.

Actually I just noticed, they don't just convert to Raw, they also convert to URL encoding - that is, no matter which base64 encoding the data uses (since it is, as of yet, allowed to vary based on use), the signature will be over the RawURL encoding of the data. I'll amend this code to make sure it does the same.

Reply:
then the special trimming logic should be executed for the interop case only, default case should be just RawURLEncoded without trimming